pkg/specs/upstream_test.go

```diff
@@ -0,0 +1,185 @@
+	tests := []struct {
+		name         string
+		upstream     string
+		currentState state.State
+		expected     string
+		expectErr    bool
+	}{
+		{
+			name:     "versioned upstream has an update available",
+			upstream: "github.com/o/r/tree/_latest_",
+			currentState: &state.VersionedState{
+				V1: &state.V1{
+					Metadata: &state.Metadata{
+						Version: "1.1.0",
+					},
+				},
+			},
+			expected: "github.com/o/r/tree/1.2.0",
+		},
+		{
+			name:     "version upstream is above latest",
+			upstream: "github.com/o/r/tree/_latest_",
+			currentState: &state.VersionedState{
+				V1: &state.V1{
+					Metadata: &state.Metadata{
+						Version: "1.2.1",
+					},
+				},
+			},
+			expected:  "",
+			expectErr: true,
+		},
+		{
+			name:     "not a versioned upstream",
+			upstream: "github.com/o/r/tree/ref",
+			currentState: &state.VersionedState{
+				V1: &state.V1{
+					Metadata: &state.Metadata{
+						Version: "1.2.0",
+					},
+				},
+			},
+			expected: "github.com/o/r/tree/ref",
+		},
+		{
+			name:     "versioned upstream with no latest release",
+			upstream: "github.com/a/b/tree/_latest_",
+			currentState: &state.VersionedState{
+				V1: &state.V1{
+					Metadata: &state.Metadata{
+						Version: "1.2.0",
+					},
+				},
+			},
+			expected:  "",
+			expectErr: true,
+		},
+		{
+			name:     "versioned upstream with no version in state",
+			upstream: "github.com/o/r/tree/_latest_",
+			currentState: &state.VersionedState{
+				V1: &state.V1{
+					Metadata: &state.Metadata{},
+				},
+			},
+			expected: "github.com/o/r/tree/1.2.0",
+		},
+	}
+
+	client, mux, _, teardown := setupGitClient()
+	mux.HandleFunc("/repos/o/r/releases/latest", func(w http.ResponseWriter, r *http.Request) {
+		latestTag := "1.2.0"
+		latest := struct {
+			TagName *string `json:"tag_name"`
+		}{
+			TagName: &latestTag,
+		}
+		w.Header().Set("Content-Type", "application/json")
+		json.NewEncoder(w).Encode(latest)
+	})
+	defer teardown()
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			testLogger := &logger.TestLogger{T: t}
+			req := require.New(t)
+			r := &Resolver{
+				GitHubFetcher: &githubclient.GithubClient{
+					Client: client,
+					Logger: testLogger,
+				},
+				Logger: testLogger,
+			}
+			actual, err := r.MaybeResolveVersionedUpstream(context.Background(), tt.upstream, tt.currentState)
+			if tt.expectErr {
+				req.Error(err)
+			} else {
+				req.NoError(err)
+			}
+			req.Equal(tt.expected, actual)
+		})
+	}
+}
+
+func TestResolver_maybeCreateVersionedUpstream(t *testing.T) {
+	tests := []struct {
+		name     string
+		upstream string
+		expected string
+	}{
+		{
+			name:     "no version - no change",
+			upstream: "github.com/helm/charts/stable/grafana",
+			expected: "github.com/helm/charts/stable/grafana",
+		},
+		{
+			name:     "has a version - changed",
+			upstream: "https://github.com/istio/istio/tree/1.0.2/install/kubernetes/helm/istio",
```

Comment:
can we add a 'ref' version of this? `https://github.com/istio/istio/tree/abcdef/install/kubernetes/helm/istio` or something similar